auto-value-gson/src/main/java/com/ryanharter/auto/value/gson/AutoValueGsonExtension.java

```
@@ -671,7 +708,166 @@ private MethodSpec createReadMethod(ClassName className,
+        readMethod.beginControlFlow("if ($N == null)", adapterField);
+        CodeBlock.Builder block = CodeBlock.builder();
+        addTypeAdapterAssignment(block, adapterField, prop, jsonAdapter, typeParams);
+        readMethod.addCode(block.build());
+        readMethod.endControlFlow();
+        block = CodeBlock.builder();
+        addBuilderFieldSetting(block, prop, builder, jsonReader, builderContext, adapters);
+        readMethod.addCode(block.build());
+        readMethod.addStatement("break");
+        readMethod.endControlFlow();
+      }
+    }
+
+    // skip value if field is not serialized...
+    readMethod.beginControlFlow("default:");
+    for (Property prop : properties) {
+      if (prop.isTransient()) {
+        continue;
+      }
+
+      if (!prop.hasSerializedNameAnnotation()) {
+        readMethod.beginControlFlow("if (realFieldNames.get($S).equals(_name))", prop.humanName);
+        FieldSpec adapterField = adapters.get(prop.type);
+        readMethod.addStatement("$T $N = this.$N", adapterField.type, adapterField, adapterField);
+        readMethod.beginControlFlow("if ($N == null)", adapterField);
+        CodeBlock.Builder block = CodeBlock.builder();
+        addTypeAdapterAssignment(block, adapterField, prop, jsonAdapter, typeParams);
+        readMethod.addCode(block.build());
+        readMethod.endControlFlow();
+        block = CodeBlock.builder();
+        addBuilderFieldSetting(block, prop, builder, jsonReader, builderContext, adapters);
+        readMethod.addCode(block.build());
+        readMethod.addStatement("continue");
+        readMethod.endControlFlow();
+      }
+    }
+    readMethod.addStatement("$N.skipValue()", jsonReader);
+    readMethod.endControlFlow(); // default case
+
+    readMethod.endControlFlow(); // switch
+    readMethod.endControlFlow(); // while
+
+    readMethod.addStatement("$N.endObject()", jsonReader);
+
+    // find the build method to use
+    if (!builderContext.buildMethod().isPresent()) {
```

Let's find this eagerly before the property code above so we can fail fast